---
title: "Oskar Expression in *Gryllus bimaculatus* tissues"
output:
  html_notebook: default
  pdf_document: default
  html_document:
    df_print: paged
---

```{r}
library(ggplot2)
library(tidyr)
library(readr)
library(stringr)
library(dplyr)
library(forcats)

knitr::opts_chunk$set(echo = TRUE, cache=TRUE)
```

## *Gryllus bimaculartus* RNA-seq data

The RNA-seq data from multiple tissues of *Gryllus bimaculatus*  was genearted by [Whittle et al. 2021](https://link.springer.com/article/10.1186/s12864-021-07411-w)

## Remove adapters from reads

Remove adapters with cutadapt v3.4.

- Bash script:

```{bash, eval=FALSE}

for R1 in $(find . -type f  -name "*.fastq.gz")
do
    prebasename=`echo $R1 | sed 's/.fastq.gz//'`
    fname=$(basename $prebasename)
    printf "\n"
        echo File:
        echo $R1
    printf "\n"
    i=$((i+1))
    echo "Sample number", $i
    echo "Sample name", $fname

     cutadapt --minimum-length 20 -o after_cutadapt/$fname.cutadapt.fastq.gz  $R1

    printf "\n"
    echo  Done with $fname 
    printf "\n"
done;

```

## Quantify gene expression with RSEM + STAR

Map reads against the genome with STAR v2.7.0e1 and quantify gene expression with rsem v1.2.29. The *G. bimaculatus* genome was generated by [Ylla et al. 2021](https://www.nature.com/articles/s42003-021-02197-9). Genome fasta and annotation (gff) files are available through [FigShare](https://figshare.com/projects/Gryllus_bimaculatus_and_Laupala_kohalensis_genome_annotations/101402).

- Index genome bash script:

```{bash, eval=FALSE}
pwd; hostname; date

# Input vars

gff3=/genome_v4/Gb_V4.onlyGenes.gff
genomefasta=/genome_v4/Gryllus_bimcaulatus_genomev4_oneline_v2.fasta

mkdir /RSEM_Index/                                   
indexname=/RSEM_Index/Gbi_V4_index

rsem-prepare-reference --gff3 $gff3 --star --num-threads $SLURM_NTASKS --star-sjdboverhang 100  $genomefasta $indexname

pwd; hostname; date

```

- Run RSEM+STAR bash script

```{bash, eval=FALSE}

pwd; hostname; date

outpputpath=/rsem_out
indexname=/RSEM_Index/Gbi_V4_index

i=0
for R1 in $(find "Fastq/after_cutadapt" -type f  -name "*.fastq.gz")
do

    prebasename=`echo $R1 | sed 's/.fastq.gz//'`
    fname=$(basename $prebasename)

    printf "\n"
        echo Mapping:
        ls -lah $R1
    printf "\n"
    i=$((i+1))
    echo "Sample number", $i
    echo "Sample name", $fname

    rsem-calculate-expression --star \
                              --phred64-quals \
                              --keep-intermediate-files \
            --num-threads $SLURM_NTASKS \
            --star-output-genome-bam \
            --star-gzipped-read-file $R1 \
                  $indexname   $outpputpath/$fname

    printf "\n"
    echo  Done with $fname 
    printf "\n"

done;

echo "All mapped!"

echo "done!"

pwd; hostname; date

```


## Load Metadata

```{r message=FALSE, warning=FALSE}
Metadata_raw<-read_csv(file = "data/Metadata_NCBI.csv")

Metadata<-Metadata_raw %>%
  mutate(Name= str_replace(LibraryName, pattern=".R1.fastq.gz", replacement="") ) %>% 
  mutate(SampleName_2= str_replace(SampleName, pattern="GB_", replacement="") ) %>% 
  mutate(SampleName_2= str_replace(SampleName_2, pattern="sample", replacement="") ) %>% 
  mutate(BioSample_GY= str_replace(SampleName_2, pattern="_[1-9]", replacement="") ) 


```




## Load gene expression

- TPMs

```{r message=FALSE, warning=FALSE}
TPM_table<-read_tsv(file = "results/RSEM_TPMs.tsv")
```

```{r}
TPM_tbl<-TPM_table %>%  
  pivot_longer(cols=colnames(TPM_table)[-1],names_to="Sample", values_to="TPMs") %>% 
  mutate(Sample= str_replace(Sample, pattern="rsem_out/", replacement="") ) %>% 
  mutate(Sample= str_replace(Sample, pattern=".R1.cutadapt.genes.results", replacement="") ) %>% 
  dplyr::rename(GeneID=X1)
```


```{r}
TPM_tbl_metadata<-TPM_tbl %>% left_join(Metadata, by=c("Sample"="Name") ) %>% 
  filter(!is.na(BioSample)) %>% #remove "Undetermined reads"  and 2 samples not in NCBI  both have low mapped reads. Perhaps are form other spp
  mutate(BioSample_GY=fct_relevel(BioSample_GY, c("embryos","female_ovary","male_Testes","female_somatic_gonad", "male_somatic_gonad","female_brain","male_brain","female_ventral_cord","female_accessory gland","male_accessory_gland","female_carcass","male_carcass")))
```



## Oskar  Expression


```{r}
Oskar_TPMs<-TPM_tbl_metadata %>% 
  filter(GeneID %in% "GBI_01840") 
```

### Expression table

```{r echo=TRUE}
Oskar_TPMs
```

- Write as csv

```{r}
Oskar_exp_TPMs<-Oskar_TPMs %>% dplyr::select(SampleName_2, TPMs)

#write_csv(Oskar_exp_TPMs, file="Oskar_Gbi_TPMs.csv")
```

- Mean per tissue

```{r echo=TRUE}
Oskar_TPMs %>% group_by(BioSample_GY) %>% summarise(mean(TPMs)) %>% arrange( `mean(TPMs)`)
```


### Plot expression

```{r, fig.width=7.2, fig.height=7.2}
Osk_exp_plot<-ggplot(Oskar_TPMs, aes(x=BioSample_GY, y= TPMs, fill=BioSample_GY), color="black")+
      geom_point( position = position_jitterdodge(jitter.width = 0.5, seed=234), size = 6, shape=21)+
      theme_bw(base_size = 14)+
      labs(title = "Expression of Oskar in Gbi tissues",y="Transcripts per Million (TPMs)",fill ="Legend",x="")+
      theme(axis.text.x = element_text(angle = 45,  size=12, hjust=1))+
      scale_fill_manual(values = rep("grey",30 )) 

Osk_exp_plot

#ggsave(Osk_exp_plot, file="Osk_exp_gbi.svg", width = 8, height = 8)
```

### Oskar aa sequences

```
>GBI_01840-RA
MSWTEVESAIKSCVVSKKGVMTLTDLEGEYYELVGQAIPFRTFGCDSLEVFLHKISGVYLSQNGRGEIVITANSKKTEHIANLVKKQKADPKALKKRTKVTYSTPKKRFWAARRNTSRHFQVFNGNVSRARASPSKSRGPGANGYSKFPADRHYAYNSRVSCTTVRYIPAGEELSDDSYGCSPDRGQKQFTSDDSDVQSDINILNVTEIGNRMIVKEEPNTSTVIAGYSAGFLNGNSSSADHYRTVTFTGDTLRCTVSGTPLPDTDYVSVNINDVEDSHNYSSWDNGPVMLGYHLVGDDFLLQFAYECLHSDYEWDDSERLTCGMCISGQTIHDFHMQVQHEGVIQPKIMLMLGAVDILQNRSLRYIIDDMTDLLDDLLISGAKTVMLLTIPFLKNAGRSNRKVLEFNEFLKTLDDGVKVQVLDVHPLFCSYYQNQTYDNCMNPKCYELSARDVLCYPNGVVLWSHIGRSYLYNFLLQQLTELENTH
>GBI_01840-RB
MSWTEVESAIKSCVVSKKGVMTLTDLEGEYYELVGQAIPFRTFGCDSLEVFLHKISGVYLSQNGRGEIVITANSKKTEHIANLVKKQKADPKALKKRTKVTYSTPKKRFWAARRNTSRHFQVFNGNVSRARASPSKSRGPGANGYSKFPADRHYAYNSRVSCTTVRYIPAGEELSDDSYGCSPDRGQKQFTSDDSDVQSDINILNVTEIGNRMIVKEEPNTSTVIAGYSAGFLNGNSSSADHYRTVTFTGDTLRCTVSGTPLPDTDYVSVNINDVEDSHNYSSWDNGPVMLGYHLVGDDFLLQFAYECLHSDYEWDDSERLTCGMCISGQTIHDFHMQVQHEGVIQPKIMLMLGAVDILQNRSLRYIIDDMTDLLDDLLISGAKTVMLLTIPFLKNAGRSNRKVLEFNEFLKTLDDGVKVQVLDVHPLFCSYYQNQTYDNCMNPKCYELSARDVLCYPNGVVLWSHIGRSYLYNFLLQQLTELENTH
```

